### PR TITLE
[Document Provider] Delete File/Folder

### PR DESCRIPTION
**Pull request by abelgardep**
Implements #2499 

____

Bugs & improvements

- [x] (1) With no connection, success message is shown https://github.com/owncloud/android/pull/2511#issuecomment-483673397

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 1]
```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
-            throw java.lang.UnsupportedOperationException("Rename failed")
+            throw FileNotFoundException("Remote Operation failed due to ${result.exception.message}")
+
         } else {
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                revokeDocumentPermission(file.fileId.toString())
+            }
             syncRequired = false
             context?.contentResolver?.notifyChange(toNotifyUri(toUri(file.parentId.toString())), null)
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: This line appears several times in the class, you might create a new method called `notifyChangeInDocument(documentId)` or something like that

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 2]
```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+
+    override fun deleteDocument(documentId: String) {
+        val docId = documentId.toLong()
+
+        updateCurrentStorageManagerIfNeeded(docId)
+
+        val file = currentStorageManager?.getFileById(docId) ?: throw FileNotFoundException("File $docId not found")
+        Log_OC.d(TAG, "Trying to delete ${file.fileName} with id ${file.fileId}")
+
+        RemoveFileOperation(file.remotePath, false).apply {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Wow, taking advantage of Kotlin power, just a couple of lines to execute the operation and check the result, love it 👏

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 3]
```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+
+        //Documents can only have one parent
+        deleteDocument(documentId)
+
+    }
+
+    private fun checkOperationResult(result: RemoteOperationResult<Any>, file: OCFile) {
 
         if (!result.isSuccess) {
+
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: I would not include this empty line here nor after `else {` but up to you

**Comment by davigonz** [comment]
@abelgardep a few changes requested

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 4]
```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+        val file = currentStorageManager?.getFileById(docId) ?: throw FileNotFoundException("File $docId not found")
+        Log_OC.d(TAG, "Trying to delete ${file.fileName} with id ${file.fileId}")
+
+        RemoveFileOperation(file.remotePath, false).apply {
+            execute(currentStorageManager, context).also { checkOperationResult(it, file) }
+        }
+    }
+
+    @TargetApi(24)
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Remove is available from API24 -> (https://developer.android.com/reference/android/provider/DocumentsContract.Document.html#FLAG_SUPPORTS_REMOVE)
Delete is available from API19 -> (https://developer.android.com/reference/android/provider/DocumentsContract.Document.html#FLAG_SUPPORTS_DELETE)

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 5]
```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
+    override fun removeDocument(documentId: String, parentDocumentId: String) {
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: It is not, is it? Files are hierarchical, so they only have one parent. I suppose this function is not needed at all.

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 1 above)
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Ok

**Comment by abelgardep** [comment]
Changes applied.

**Comment by jesmrec** [comment]
Let's QA

**Comment by jesmrec** [comment]
### (1)

1. Remove the connection of the device, stop the server or set the maintenance mode
2. Try to delete content using document provider

Current: File is not deleted, but snackbar shows success message
Expect: Some kind of error, since the file was not deleted. Is this something feasible or is it handled directly by Android?

Android 8

**Comment by abelgardep** [comment]
@jesmrec 

In those cases, in Android 7.1.1 appears a notification error: 'These files weren't deleted: [List of files]'

**Comment by jesmrec** [comment]
The matter was in the notification. There, we can see correctly that the files were not deleted.

**Comment by jesmrec** [comment]
Approved.

**Pull request merged**